core/primitives/src/state.rs

```diff
@@ -0,0 +1,60 @@
+use byteorder::{LittleEndian, ReadBytesExt};
+use near_primitives_core::hash::{hash, CryptoHash};
+use std::io::{Cursor, Read};
+
+// to specify a part we always specify both part_id and num_parts together
+#[derive(Copy, Clone)]
+pub struct PartId {
+    pub idx: u64,
+    pub total: u64,
+}
+impl PartId {
+    pub fn new(part_id: u64, num_parts: u64) -> PartId {
+        assert!(part_id < num_parts);
+        PartId { idx: part_id, total: num_parts }
+    }
+}
+
+/// State value reference. Used to charge fees for value length before retrieving the value itself.
+pub struct ValueRef {
+    /// Value length in bytes.
+    pub length: u32,
+    /// Unique value hash.
+    pub hash: CryptoHash,
+}
+
+impl ValueRef {
+    /// Create serialized value reference by the value.
+    /// Resulting array stores 4 bytes of length and then 32 bytes of hash.
+    pub fn create_serialized(value: &[u8]) -> [u8; 36] {
+        let mut result = [0u8; 36];
+        result[0..4].copy_from_slice(&(value.len() as u32).to_le_bytes());
+        result[4..36].copy_from_slice(&hash(value).0);
+        result
+    }
+
+    /// Decode value reference from the raw byte array.
+    pub fn decode(bytes: &[u8]) -> Result<Option<Self>, std::io::Error> {
```

Comment:
- I don't think I understand why we need to return an `Option<Self>` here.  I don't see any place where we return `None`.
- I know we use this pattern in other places but if we were to not use `Cursor` and instead return an error if the length of `bytes` is not 36, then we will only be throwing a single type of error instead of the generic `std::io::Error` which is very difficult to handle.

Reply:
Fixed return type, adding TODO for fixing the error handling - it is the same in the Trie code.